module/zfs/arc.c

```diff
@@ -6166,7 +6166,27 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 			    metadata, misses);
 		}
 
-		if (vd != NULL && l2arc_ndev != 0 && !(l2arc_norw && devw)) {
+		/*
+		 * Check if the spa even has l2 configured;
+		 *
+		 * NOTE #1: not holding spa_namespace_lock -
```

Comment:
We shouldn't need holding spa_namespace_lock because we are not making any changes to the pool configuration here. 
We lock out L2ARC device removal with spa_config_tryenter() a few lines above. I think you can remove NOTE#1 entirely.

Reply:
Good - I was hoping that was the case, the NOTEs were articulate my concerns so I could be corrected. :D